src/libraries/System.Private.CoreLib/src/System/Math.DivModInt.cs

```diff
@@ -0,0 +1,234 @@
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if (Hi32Bits(dividend) == 0)
+                    return (uint)dividend / (uint)divisor;
+            }
+
+            return DivUInt64Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static int ModInt32(int dividend, int divisor)
+        {
+            if ((uint)(divisor + 1) <= 1) // Unsigned test for divisor in [-1 .. 0]
+            {
+                if (divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+                else if (divisor == -1)
+                {
+                    if (dividend == int.MinValue)
+                    {
+                        ThrowHelper.ThrowOverflowException();
+                    }
+                    return 0;
+                }
+            }
+
+            return ModInt32Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static uint ModUInt32(uint dividend, uint divisor)
+        {
+            if (divisor == 0)
+            {
+                ThrowHelper.ThrowDivideByZeroException();
+            }
+
+            return ModUInt32Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static long ModInt64(long dividend, long divisor)
+        {
+            if (Is32BitSigned(divisor))
+            {
+                if ((int)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if ((int)divisor == -1)
+                {
+                    if ((ulong)dividend == 0x8000000000000000ul)
+                    {
+                        ThrowHelper.ThrowOverflowException();
+                    }
+                    return 0;
+                }
+
+                if (Is32BitSigned(dividend))
+                {
+                    return (int)dividend % (int)divisor;
+                }
+            }
+
+            return ModInt64Internal(dividend, divisor);
+        }
+
+        [StackTraceHidden]
+        internal static ulong ModUInt64(ulong dividend, ulong divisor)
+        {
+            if (Hi32Bits(divisor) == 0)
+            {
+                if ((uint)divisor == 0)
+                {
+                    ThrowHelper.ThrowDivideByZeroException();
+                }
+
+                if (Hi32Bits(dividend) == 0)
+                    return (uint)dividend % (uint)divisor;
+            }
+
+            return ModUInt64Internal(dividend, divisor);
+        }
+
+        // helper method to get high 32-bit of 64-bit int
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint Hi32Bits<T>(T a) where T : INumber<T>
+        {
+            return (uint)(ulong.CreateTruncating(a) >> 32);
+        }
+
+        // helper method to check whether 64-bit signed int fits into 32-bit signed (compiles into one 32-bit compare)
```

Comment:
If we go with the plan to have the JIT to inline the helpers as IL, I do not think we want to have these checks inlined to make it less work for the JIT to deal with inlining them.